helix-core/src/main/java/org/apache/helix/controller/rebalancer/topology/Topology.java

```diff
@@ -212,123 +186,106 @@ private static Node cloneTree(Node root, Map<Node, Integer> newNodeWeight, Set<N
     return newRoot;
   }
 
-  /**
-   * Creates a tree representing the cluster structure using default cluster topology definition
-   * (i,e no topology definition given and no domain id set).
-   */
-  private Node createClusterTreeWithDefaultTopologyDef() {
+  private Node createClusterTree() {
     // root
     Node root = new Node();
     root.setName("root");
     root.setId(computeId("root"));
     root.setType(Types.ROOT.name());
 
-    for (String ins : _allInstances) {
-      InstanceConfig config = _instanceConfigMap.get(ins);
-      Map<String, String> pathValueMap = new HashMap<>();
-      if (_topologyAwareEnabled) {
-        String zone = config.getZoneId();
+    for (String instance : _allInstances) {
+      InstanceConfig insConfig = _instanceConfigMap.get(instance);
+      LinkedHashMap<String, String> instanceTopologyMap = new LinkedHashMap<>();
```

Comment:
@xyuanlu 
Let's use a Map to declare?

And what you said about LinkedHashMap is true, but Map is a higher-level construct (interface) and LinkedHashMap is an implementation. Even if you wanted a LinkedHashMap instance, how do you guarantee that the map has the right order? If you want to be 100% sure about the order, you have two options:

1. Make it crystal clear that the user must pass in ordered map (TreeMap or LinkedHashMap)
2. do your own sorting internally.

With that said, what you said about using LinkedHashMap for declaration implies that maybe we should review the difference between declaration and implementation. Here's a helpful explanation: https://stackoverflow.com/questions/11715485/what-is-the-difference-between-declaration-and-definition-in-java

Reply:
TFTR.
Please correct me if I am wrong. As a general rule, we're supposed use its interface in declaration rather than the actual implementation like the following.
`Map<String, String> instanceTopologyMap = new LinkedHashMap<>();`
The reason behind this is that the implementation of these interfaces is usually not relevant when handling them in the following logic. However, in my understanding, in some special cases like this, if we need a map when order is important, it is better to use a more specific interface or an implementation to explicitly state that the order is important. Or maybe still use the general interface with comment?

Reply:
Yes, what you said about using the interface is on point.
I understand where you are coming from - you want to make it explicit that the parameter that's being passed in is ordered. But let's revisit what I said above - making the declaration LinkedHashMap does not help you in that case.

LinkedHashMap's ordering is not deterministic. It does not enforce the right ordering you need in the map. So you aren't really enforcing anything there other than the map will have some order.

If you want to enforce some sort of ordering of elements, you should sort internally. Say you need it sorted by alphabetical order, then you should convert this map into a TreeMap. 

Alternatively, stating that the topology map must be ordered in some way is also acceptable in JavaDoc, but in that case, the right thing to do is to be able to detect if the ordering has a problem, be able to throw an exception of some sort, alerting the user.

Reply:
For this in class variable defined here in a private function for class constructor, i guess it is safe to assume the order will be maintained as expected  since this private function has the full control of this object. Of course, more detailed comment/doc should be added.

I think it is safe to use the interface map to declare a in class object here. The public util function(computeInstanceTopologyMap) however, takes this LinkedHashMap instance, populate it and return as an out param.